### ISSUE TITLE
Testing for errors

### ISSUE THREAD

**Issue by oysteins** [fingolfin/UnitTests#2]
It should be possible to write tests for cases where the expected behaviour is that an error is signalled.

I suggest adding a function `AssertError` which takes three arguments: a string `msg`, a function `f` and a list `args`.  The call

```
AssertError( msg, f, args )
```

asserts that applying the function `f` to the argument list `args` produces an error whose error message starts with the string `msg`.

The following is a suggestion for an implementation of this function:

``` GAP
InstallGlobalFunction( AssertError, function(msg, f, args)
  local OriginalErrorFunction, res, errmsg;
  OriginalErrorFunction := Error;
  MakeReadWriteGlobal("Error");
  Error := function(arg)
    local errmsg;
    errmsg := Concatenation( List( arg, String ) );
    JUMP_TO_CATCH( errmsg );
  end;
  MakeReadOnlyGlobal("Error");
  res := CALL_WITH_CATCH( f, args );
  MakeReadWriteGlobal("Error");
  Error := OriginalErrorFunction;
  MakeReadOnlyGlobal("Error");
  if res[ 1 ] then
    JUMP_TO_CATCH( Concatenation( "expected Error(", msg, "...) but got no error" ) );
  else
    errmsg := res[ 2 ];
    if not StartsWith( errmsg, msg ) then
      JUMP_TO_CATCH( Concatenation( "expected Error(", msg, "...) but got Error(", errmsg, ")" ) );
    fi;
  fi;
end );
```

A silly example demonstrating use of `AssertError`:

``` GAP
f := function( x )
  if x = 42 then
    return true;
  else
    Error( "Wrong answer: ", x );
  fi;
end;
AssertError( "Wrong answer", f, [ 5 ] );
AssertError( "Wrong answer", f, [ 42 ] );
AssertError( "Something completely different", f, [ 5 ] );
```

In this example, the first assertion succeeds, since the call `f( 5 )` does indeed signal an error whose message starts with "Wrong answer".  The second assertion fails, since `f( 42 )` does not signal an error.  The third assertion also fails, this time since the error message we get from `f( 5 )` does not match the prefix "Something completely different".

**Comment by oysteins** [comment]
A comment about the first argument to `AssertError`: Why do I want this to be a prefix of the error message, instead of being equal to the error message?  Because I believe this makes it easier to write tests, and possibly makes the tests more robust.

The actual error messages that are produced by a program might be long and contain some detail about the objects involved in causing the error.  Constructing the exact same message in the test requires a bit of work, and it would be a pain to have to rewrite tests just because of some small change in the appearance of an error message.

Furthermore, including just the first few words of the error message should usually be enough to distinguish between different errors.  For me, the whole point of the `msg` argument is just to make sure (with reasonable certainty) that we get the error we expect, and not a completely unrelated error.

That being said, I think it would be even better if the argument `msg` had been a regular expression which the error message should match.  Does there exist some GAP package which provides regexp functionality?

**Comment by oysteins** [comment]
The implementation I suggested above for `AssertError` takes care of testing for expected errors, but leaves the problem of unexpected errors unresolved.

To catch unexpected errors as well, one possibility would be to take the code which changes the `Error` function out from `AssertError`, and put it into `RunTestSuite` instead.  This would then replace the code in `RunTestSuite` which modifies the `BreakOnError` variable.  Then `RunTestSuite` will behave like this: for every test, we first replace the `Error` function with our version, then run the test using `CALL_WITH_CATCH`, then restore the original `Error` function.

Now, this means that there are two different kinds of situations in which we interrupt the control flow by calling `JUMP_TO_CATCH`: an assertion failure or an error.  To distinguish these, we could for example use a list where the first element indicates the reason why the program was interrupted, and the second element is the message string.  So an assertion failure would use code like

```
JUMP_TO_CATCH( [ "assertion", "expected true but got false" ] );
```

while the `Error` function would use

```
JUMP_TO_CATCH( [ "error", errmsg ] );
```

This (along with some small modifications to the code which prints out the results) should handle all errors nicely.

Well, almost.  It seems like some builtin GAP functions signal errors without invoking the `Error` function.  This is for example the case with `Remove`, as evidenced by the following GAP session:

```
gap> MakeReadWriteGlobal( "Error" );
gap> Error := function( arg ) Print( "This is my new error function\n" ); end;;
gap> Error();
This is my new error function
gap> Remove([]);
Error, Remove: <list> must not be empty
not in any function at line 4 of *stdin*
you may replace <list> via 'return <list>;'
brk>
```

Clearly, `Remove` signals errors in some way which does not involve looking up the global variable `Error` and calling that function.  What makes this even more confusing is that if we try to investigate the matter, GAP will happily tell us that `Remove` is an operation and that the method chosen in this case actually does call `Error`:

```
gap> Remove;
<Operation "Remove">
gap> ApplicableMethod( Remove, [ [] ], 6 );
#I  Searching Method for Remove with 1 arguments:
#I  Total: 2 entries
#I  Method 1: ``Remove: for a plist matrix'', value: 16
#I   - 1st argument needs [ "IsPlistMatrixRep" ]
#I  Method 2: ``Remove: one argument'', value: 4
#I  Function Body:
function ( l )
    local  x, len;
    len := Length( l );
    if len = 0  then
        Error( "Remove: list <l> must not be empty.\n" );
    fi;
    x := l[len];
    Unbind( l[len] );
    return x;
endfunction( l ) ... end
```

This is, however, all a lie.  Apparently, `Remove` is not really an operation (although GAP claims it is), and when we invoke it, GAP for some reason bypasses the method selection completely and instead calls some C code (the function `RemList` in `listfunc.c`), which in turn signals an error through some other C code (the function `ErrorReturnObj` in `gap.c`).

Continuing from there, we get to the C function `CallErrorInner`, which actually invokes the GAP function `ErrorInner`!  So maybe the trick is to overwrite the variable `ErrorInner` rather than `Error`.  Side remark: When trying this out, I learned that experimenting with the value of `ErrorInner` is a great way to crash GAP:

```
gap> MakeReadWriteGlobal("ErrorInner");
gap> ErrorInner := function( arg ) Print( "Hello, this is my new ErrorInner function\n" ); end;;
gap> Remove([]);
Hello, this is my new ErrorInner function
Segmentation fault
```

Anyway, using this as a replacement for `ErrorInner` seems to work:

```
ErrorInner := function(arg)
  local errmsg;
  errmsg := Concatenation( List( arg[ 2 ], String ) );
  JUMP_TO_CATCH( [ "error", errmsg ] );
end;
```

To sum up, I suggest the following:
- In `RunTestSuite`, overwrite `ErrorInner` with a new function (as described above) instead of setting `BreakOnError` (and then restore the original value of `ErrorInner` after calling the test).
- Modify the assert functions to give a list `[ "assertion", message ]` to `JUMP_TO_CATCH` instead of just the message, and modify the code that receives the messages accordingly.
- Include `AssertError` as described earlier, but without the code that changes `Error` (it will instead rely on the fact that `ErrorInner` has been modified).
